### PR TITLE
Add ability to obtain Steam Auth Tickets

### PR DESCRIPTION
**Pull request by clydejallorina**
This PR is for adding the ability to send over Steam Tickets on Signup to link Steam accounts with incoming TootTally accounts. This would be used in verifying Steam users and simplifying the login process by allowing Steam users to login by their Steam Auth Ticket.

### CODE REVIEW

**Comment by clydejallorina** [comment]
Bump

File: Utils/Steam/SteamAuthTicketHandler.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                m_GetAuthSessionTicketResponse = Callback<GetTicketForWebApiResponse_t>.Create(OnSteamAuthTicketResponse);
+                hAuthTicket = SteamUser.GetAuthTicketForWebApi("TootTally");
+            }
+            else
+            {
+                Plugin.LogWarning("SteamManager not initialized, cannot obtain Steam Auth ticket.");
+            }
+        }
+
+        public static void Cleanup()
```
Comment (hedgehog1029) [inline Utils/Steam/SteamAuthTicketHandler.cs:43]: I don't see this getting called anywhere, is it?

**Review by hedgehog1029 — APPROVED** [review]

**Comment by clydejallorina** [comment]
git diff since the rebase kinda screwed over the Files Changed tab :/
```diff
diff --git a/APIService/SerializableClass.cs b/APIService/SerializableClass.cs
index 073c618..4c03d49 100644
--- a/APIService/SerializableClass.cs
+++ b/APIService/SerializableClass.cs
@@ -113,6 +113,7 @@ namespace TootTallyCore.APIServices
         public class ModInfoAPI
         {
             public string apiKey;
+            public string steamTicket;
             public SendableModInfo[] mods;
         }
 
@@ -160,12 +161,20 @@ namespace TootTallyCore.APIServices
             public int status;
         }
 
+        [Serializable]
+        public struct APISteamConnect
+        {
+            public string apiKey;
+            public string steamTicket;
+        }
+
         [Serializable]
         public class APISignUp
         {
             public string username;
             public string password;
             public string pass_check;
+            public string steamTicket;
         }
 
         [Serializable]
@@ -175,6 +184,12 @@ namespace TootTallyCore.APIServices
             public string password;
         }
 
+        [Serializable]
+        public class APISteamLogin
+        {
+            public string steamTicket;
+        }
+
         [Serializable]
         public class LoginToken
         {
diff --git a/APIService/TootTallyAPIService.cs b/APIService/TootTallyAPIService.cs
index 47ef18d..62ee080 100644
--- a/APIService/TootTallyAPIService.cs
+++ b/APIService/TootTallyAPIService.cs
@@ -7,6 +7,7 @@ using Newtonsoft.Json;
 using TootTallyCore.Graphics;
 using TootTallyCore.Utils.Helpers;
 using TootTallyCore.Utils.TootTallyNotifs;
+using TootTallyCore.Utils.Steam;
 using UnityEngine;
 using UnityEngine.Networking;
 using static TootTallyCore.APIServices.SerializableClass;
@@ -111,6 +112,35 @@ namespace TootTallyCore.APIServices
                 callback(null);
         }
 
+        public static IEnumerator<UnityWebRequestAsyncOperation> GetUserFromSteamTicket(Action<User> callback)
+        {
+            var query = $"{APIURL}/auth/steam-login/";
+            var steamTicket = SteamAuthTicketHandler.SteamTicket;
+            var apiObj = new APISteamLogin() { steamTicket = steamTicket };
+            var apiLogin = System.Text.Encoding.UTF8.GetBytes(JsonUtility.ToJson(apiObj));
+            var webRequest = PostUploadRequest(query, apiLogin);
+            User user;
+
+            yield return webRequest.SendWebRequest();
+
+            if (!HasError(webRequest, query))
+            {
+                user = JsonConvert.DeserializeObject<User>(webRequest.downloadHandler.text);
+                Plugin.LogInfo($"Welcome, {user.username}!");
+            }
+            else
+            {
+                user = new User()
+                {
+                    username = "Guest",
+                    id = 0,
+                };
+                Plugin.LogInfo($"Logged in with Guest Account");
+            }
+
+            callback(user);
+        }
+
         public static IEnumerator<UnityWebRequestAsyncOperation> GetLoginToken(string username, string password, Action<LoginToken> callback)
         {
             var query = $"{APIURL}/auth/token/";
@@ -138,7 +168,8 @@ namespace TootTallyCore.APIServices
         public static IEnumerator<UnityWebRequestAsyncOperation> SignUpRequest(string username, string password, string pass_check, Action<bool> callback)
         {
             var query = $"{APIURL}/auth/signup/";
-            var apiObj = new APISignUp() { username = username, password = password, pass_check = pass_check };
+            var steamTicket = SteamAuthTicketHandler.SteamTicket;
+            var apiObj = new APISignUp() { username = username, password = password, pass_check = pass_check, steamTicket = steamTicket };
             var apiSignUp = System.Text.Encoding.UTF8.GetBytes(JsonUtility.ToJson(apiObj));
             var webRequest = PostUploadRequest(query, apiSignUp);
             yield return webRequest.SendWebRequest();
@@ -406,6 +437,7 @@ namespace TootTallyCore.APIServices
             }
 
             sendableModInfo.apiKey = apiKey;
+            sendableModInfo.steamTicket = SteamAuthTicketHandler.SteamTicket;
             sendableModInfo.mods = mods.ToArray();
             string query = $"{APIURL}/api/mods/submit/";
             var jsonbin = System.Text.Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(sendableModInfo));
@@ -433,6 +465,23 @@ namespace TootTallyCore.APIServices
                 callback?.Invoke();
         }
 
+        public static IEnumerator<UnityWebRequestAsyncOperation> ConnectSteamToProfile(string apiKey, int status, Action callback = null)
+        {
+            APISteamConnect steamConnect = new APISteamConnect()
+            {
+                apiKey = apiKey,
+                steamTicket = SteamAuthTicketHandler.SteamTicket,
+            };
+
+            string query = $"{APIURL}/api/profile/connect_steam/";
+            var data = System.Text.Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(steamConnect));
+            UnityWebRequest webRequest = PostUploadRequest(query, data);
+            yield return webRequest.SendWebRequest();
+
+            if (!HasError(webRequest, query))
+                callback?.Invoke();
+        }
+
         public static IEnumerator<UnityWebRequestAsyncOperation> GetLatestOnlineUsers(int id, Action<List<User>> callback)
         {
             string query = $"{APIURL}/api/users/latest/?userID={id}";
diff --git a/Plugin.cs b/Plugin.cs
index 80eae47..b6026b7 100644
--- a/Plugin.cs
+++ b/Plugin.cs
@@ -12,6 +12,7 @@ using TootTallyCore.Utils.Assets;
 using TootTallyCore.Utils.TootTallyGlobals;
 using TootTallyCore.Utils.TootTallyModules;
 using TootTallyCore.Utils.TootTallyNotifs;
+using TootTallyCore.Utils.Steam;
 using UnityEngine;
 
 namespace TootTallyCore
@@ -108,6 +109,8 @@ namespace TootTallyCore
             gameObject.AddComponent<TootTallyAnimationManager>();
 
             TootTallyModuleManager.LoadModules();
+            SteamAuthTicketHandler.GetSteamAuthTicket();
+
             LogInfo($"Plugin {PluginInfo.PLUGIN_GUID} [Build {BUILDDATE}] is loaded!");
             LogInfo($"Game Version: {Application.version}");
         }
diff --git a/TootTallyCore.csproj b/TootTallyCore.csproj
index bbaca2c..f437b57 100644
--- a/TootTallyCore.csproj
+++ b/TootTallyCore.csproj
@@ -28,9 +28,9 @@
     <PackageReference Include="BepInEx.Analyzers" Version="1.*" PrivateAssets="all" />
     <PackageReference Include="BepInEx.Core" Version="5.*" />
     <PackageReference Include="BepInEx.PluginInfoProps" Version="1.*" />
-    <PackageReference Include="TromboneChamp.GameLibs" Version="1.20.0-beta" />
-    <PackageReference Include="TromboneChamp.BaboonAPI" Version="2.7.0-alpha.1" />
-    <PackageReference Include="TromboneChamp.TrombLoader" Version="2.0.0-rc.1" />
+    <PackageReference Include="TromboneChamp.GameLibs" Version="1.22.0" />
+    <PackageReference Include="TromboneChamp.BaboonAPI" Version="2.7.3" />
+    <PackageReference Include="TromboneChamp.TrombLoader" Version="2.3.0" />
     <PackageReference Include="UnityEngine.Modules" Version="2019.4.40" IncludeAssets="compile" />
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
     <PackageReference Include="websocket-sharp-customheaders" Version="1.0.2.31869" />
diff --git a/Utils/Helpers/HexConverter.cs b/Utils/Helpers/HexConverter.cs
new file mode 100644
index 0000000..a13ba96
--- /dev/null
+++ b/Utils/Helpers/HexConverter.cs
@@ -0,0 +1,33 @@
+∩╗┐using System;
+
+namespace TootTallyCore.Utils.Helpers
+{
+    public class HexConverter
+    {
+        private static readonly uint[] Lookup32 = CreateLookup32();
+
+        private static uint[] CreateLookup32()
+        {
+            var result = new uint[256];
+            for (int i = 0; i < 256; i++)
+            {
+                string s = i.ToString("X2");
+                result[i] = ((uint)s[0]) + ((uint)s[1] << 16);
+            }
+            return result;
+        }
+
+        public static string ToHexString(Span<byte> bytes)
+        {
+            var lookup32 = Lookup32;
+            var result = new char[bytes.Length * 2];
+            for (int i = 0; i < bytes.Length; i++)
+            {
+                var val = lookup32[bytes[i]];
+                result[2*i] = (char)val;
+                result[2*i + 1] = (char) (val >> 16);
+            }
+            return new string(result);
+        }
+    }
+}
diff --git a/Utils/Steam/SteamAuthTicketHandler.cs b/Utils/Steam/SteamAuthTicketHandler.cs
new file mode 100644
index 0000000..fc1ae90
--- /dev/null
+++ b/Utils/Steam/SteamAuthTicketHandler.cs
@@ -0,0 +1,43 @@
+using Steamworks;
+using System;
+using TootTallyCore.Utils.Helpers;
+
+namespace TootTallyCore.Utils.Steam
+{
+    public static class SteamAuthTicketHandler
+    {
+        public static string SteamTicket { get; private set; }
+        private static Callback<GetTicketForWebApiResponse_t> m_GetAuthSessionTicketResponse;
+        private static HAuthTicket hAuthTicket;
+
+        private static void OnSteamAuthTicketResponse(GetTicketForWebApiResponse_t pCallback)
+        {
+            if (pCallback.m_eResult != EResult.k_EResultOK)
+            {
+                Plugin.LogError($"Unable to get Steam auth ticket (Steam API Result Code {pCallback.m_eResult})");
+            }
+            if (pCallback.m_hAuthTicket == hAuthTicket)
+            {
+                var ticketBytes = new Span<byte>(pCallback.m_rgubTicket, 0, pCallback.m_cubTicket);
+                var hex = HexConverter.ToHexString(ticketBytes);
+
+                SteamTicket = hex;
+
+                Plugin.LogInfo("Steam Auth Ticket obtained successfully!");
+            }
+        }
+
+        public static void GetSteamAuthTicket()
+        {
+            if (SteamManager.Initialized)
+            {
+                m_GetAuthSessionTicketResponse = Callback<GetTicketForWebApiResponse_t>.Create(OnSteamAuthTicketResponse);
+                hAuthTicket = SteamUser.GetAuthTicketForWebApi("TootTally");
+            }
+            else
+            {
+                Plugin.LogWarning("SteamManager not initialized, cannot obtain Steam Auth ticket.");
+            }
+        }
+    }
+}
```

**Comment by Electrostats** [comment]
Reviewed, rest looks good

**Pull request merged**